packages/block-editor/src/components/block-list/use-block-props/use-block-refs.js

```diff
@@ -67,7 +73,17 @@ function useBlockRef( clientId ) {
  */
 function useBlockElement( clientId ) {
 	const { refsMap } = useContext( BlockRefs );
-	return useObservableValue( refsMap, clientId ) ?? null;
+	const [ blockElement, setBlockElement ] = useState( null );
+	// Delay setting the resulting `blockElement` until an effect. If the block element
+	// changes (i.e., the block is unmounted and re-mounted), this allows enough time
+	// for the ref callbacks to clean up the old element and set the new one.
+	useLayoutEffect( () => {
+		setBlockElement( refsMap.get( clientId ) );
+		return refsMap.subscribe( clientId, () =>
+			setBlockElement( refsMap.get( clientId ) )
```

Comment:
On unmount, shouldn't it be set to `null`?

Reply:
I was considering this, but I think it's not needed, it wouldn't have any effect. On unmount, the component is going away, and updating its local state is not needed, because nobody is going to read it.

But maybe we can add it for consistency sake. So that the behavior is indistinguishable from the classic "store ref as state" pattern:
```js
const [ el, setEl ] = useState( null );
return <div ref={ setEl } />;
```
In this case the `setEl` will be called with `null` on unmount.

`useBlockRefs` is basically the same pattern, store an element ref into state, but the ref and the state are in different components, and the store in the provider serves as a teleportation machine between them.

Reply:
Wouldn't it be possible to for a block to disappear, while another component is using `useBlockElement`, which would then still point to an unmounted DOM node?

Reply:
Before the block disappears, the ref callback is called and sets the ref to `null`. Then the observable map calls its listeners and inside the `useBlockElement` hook, the `setBlockElement` inside the listener is called, and sets the local state to `null`. Therefore, all the references to the unmounted DOM node are cleared.

Reply:
This issue makes me thinking that our current API for reading the elements is not very good. Ideally it should work exactly like element refs. Like this:
```jsx
const blockElRef = useRef(); // can also be a ref callback
useBlockElementRef( clientId, blockElRef );
```
The `blockElRef` would be set and unset in real time as the target element changes. The point is that it is identical to how refs to local elements work:
```jsx
const blockElRef = useRef(); // can also be a ref callback
return <div ref={ blockElRef } />;
```
Exactly the same thing, only in one case the element is local and in the other it's teleported from another location.

This is much better than the current `useBlockRef` hook that _returns_ a ref with a synthetic `ref.current` getter.

The `useBlockElementRef` primitive can be used to very easily implement `useBlockElement` that returns the element and triggers a rerender on change:
```jsx
function useBlockElement( clientId ) {
  const [ el, setEl ] = useState( null );
  useBlockElementRef( clientId, setEl );
  return el;
}
```
This implementation also never has the bug with returning stale unmounted elements.

Reply:
Oh, sorry, I misread this as a change to the hook inside the block rather than a change to the hook in the consumer.

Yes, that alternative seems good to me. We should try it :)

Reply:
Implemented in #63799 🙂